src/interface/swarm.cpp

```diff
@@ -343,6 +450,360 @@ void Swarm::Defrag() {
 
   // Update max_active_index_
   max_active_index_ = num_active_ - 1;
+
+  GetBlockPointer()->exec_space.fence();
+}
+
+void Swarm::SetupPersistentMPI() {
+  printf("[%i] SetupPersistentMPI\n", Globals::my_rank);
+  vbvar->SetupPersistentMPI();
+
+  allreduce_request_ = MPI_REQUEST_NULL;
+
+  // Index into neighbor blocks
+  auto pmb = GetBlockPointer();
+  auto neighborIndices_h = neighborIndices_.GetHostMirror();
+
+  // Region belonging to this meshblock
+  // for (int k = 1; k < 3; k++) {
+  for (int k = 0; k < 4; k++) { // 2D ONLY!
+    for (int j = 1; j < 3; j++) {
+      for (int i = 1; i < 3; i++) {
+        neighborIndices_h(k, j, i) = -1;
+      }
+    }
+  }
+
+  int ndim = pmb->pmy_mesh->ndim;
+
+  PARTHENON_REQUIRE(ndim == 2, "Only 2D tested right now!");
+
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+    PARTHENON_REQUIRE(pmb->loc.level == nb.snb.level,
+                      "Particles do not currently support AMR!");
+
+    int i = nb.ni.ox1;
+    int j = nb.ni.ox2;
+    int k = nb.ni.ox3;
+
+    if (ndim == 1) {
+      if (i == -1) {
+        neighborIndices_h(0, 0, 0) = n;
+      } else if (i == 0) {
+        neighborIndices_h(0, 0, 1) = n;
+        neighborIndices_h(0, 0, 2) = n;
+      } else {
+        neighborIndices_h(0, 0, 3) = n;
+      }
+    } else if (ndim == 2) {
+      if (i == -1) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 0) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 0) = n;
+          neighborIndices_h(0, 2, 0) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 0) = n;
+        }
+      } else if (i == 0) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 1) = n;
+          neighborIndices_h(0, 0, 2) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 1) = n;
+          neighborIndices_h(0, 3, 2) = n;
+        }
+      } else if (i == 1) {
+        if (j == -1) {
+          neighborIndices_h(0, 0, 3) = n;
+        } else if (j == 0) {
+          neighborIndices_h(0, 1, 3) = n;
+          neighborIndices_h(0, 2, 3) = n;
+        } else if (j == 1) {
+          neighborIndices_h(0, 3, 3) = n;
+        }
+      }
+    } else {
+      PARTHENON_FAIL("3D particles not currently supported!");
+    }
+  }
+
+  neighborIndices_.DeepCopy(neighborIndices_h);
+}
```

Comment:
I agree but I struggled to do this in a smarter way... maybe others see a better path forward.

Reply:
I would just suggest breaking up the index setting into subroutines maybe?

Reply:
I added 3D and it's even worse and doesn't even cover AMR :sob: 

That being said I think this will just need to be rethought to get AMR working, and my preference is to do that in a separate PR. I can add e.g. `Swarm::SetNeighborIndices1D_` etc. if you think that will help in the meantime